src/ast/document-descriptor.ts

```diff
@@ -234,4 +264,30 @@ export class Document implements Feature {
 
     return result;
   }
+
+  private _featuresByKind: Map<string, Set<Feature>> = null;
+  private _featuresByKindAndId: Map<string, Map<string, Set<Feature>>> = null;
+  private _initIndexes() {
```

Comment:
If the indices are caches, this could be named `_prepoulateCaches` or `_warmupCaches` or something.